### PR TITLE
Suricata XDP Syncookie

### PR DESCRIPTION
**Pull request by vincentmli**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Describe changes:
- Add XDP Syncookie feature to prevent host from SYN flooding attack in AF_PACKET IDS

### Provide values to any of the below to override the defaults.

To use a pull request use a branch name like `pr/N` where `N` is the
pull request number.

Alternatively, `SV_BRANCH` may also be a link to an
OISF/suricata-verify pull-request.

```
SV_REPO=
SV_BRANCH=
SU_REPO=
SU_BRANCH=
LIBHTP_REPO=
LIBHTP_BRANCH=
```

### CODE REVIEW

**Comment by vincentmli** [comment]
Hi @victorjulien

I send this XDP syncookie PR and noticed the build failed on Debian 11 for the XDP syncookie program and user space program, I suspect Debian 11 may have old libbpf and kernel that does not support the XDP program and user space program. (https://github.com/OISF/suricata/actions/runs/7699348152/job/20980796747?pr=10268).

the XDP kernel program contains bpf kernel function that is supported from kernel 6.2 and above, I suspect none of existing Suricata distribution CI test has 6.2 kernel, Could you share your suggestion on this?

File: ebpf/Makefile.am [hunk 1]
```diff
@@ -1,31 +1,39 @@
...
 
+xdp_synproxy: xdp_synproxy.c
+	$(CLANG) -static $(CLANG_FLAGS) -o $@ $< $(CLANG_LFLAGS)
 
 $(BPF_TARGETS): %.bpf: %.c
 #      From C-code to LLVM-IR format suffix .ll (clang -S -emit-llvm)
 	${CLANG} -Wall $(BPF_CFLAGS) -O2 -g \
 		-I/usr/include/$(build_cpu)-$(build_os)/ \
 		-D__KERNEL__ -D__ASM_SYSREG_H \
+		-Wno-unused-value \
```
Comment (victorjulien) [inline ebpf/Makefile.am:29]: can we avoid these in favor of addressing the warnings?

File: ebpf/include/vmlinux/vmlinux_common.h [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#ifndef __VMLINUX_COMMON_H__
```
Comment (victorjulien) [inline ebpf/include/vmlinux/vmlinux_common.h:1]: what is the origin of this file? similar for the other headers below

File: ebpf/xdp_synproxy.c [hunk 3]
```diff
@@ -0,0 +1,283 @@
+// SPDX-License-Identifier: LGPL-2.1 OR BSD-2-Clause
```
Comment (victorjulien) [inline ebpf/xdp_synproxy.c:1]: where does this come from? Did you modify it?

File: ebpf/xdp_synproxy_kern.c [hunk 4]
```diff
@@ -0,0 +1,799 @@
+// SPDX-License-Identifier: LGPL-2.1 OR BSD-2-Clause
```
Comment (victorjulien) [inline ebpf/xdp_synproxy_kern.c:1]: these files look like they are pulled in from somewhere. Where are they from? Did you modify them? Is it necessary to have them in our repo?

File: src/detect-byte-extract.c [hunk 5]
```diff
@@ -629,7 +629,7 @@ static int DetectByteExtractSetup(DetectEngineCtx *de_ctx, Signature *s, const c
  error:
     DetectByteExtractFree(de_ctx, data);
     return ret;
-}
```
Comment (victorjulien) [inline src/detect-byte-extract.c:632]: unrelated change, also it's wrong

File: src/detect-bytejump.c [hunk 6]
```diff
@@ -599,7 +599,7 @@ static int DetectBytejumpSetup(DetectEngineCtx *de_ctx, Signature *s, const char
     }
     DetectBytejumpFree(de_ctx, data);
     return ret;
-}
+ }
```
Comment (victorjulien) [inline src/detect-bytejump.c:602]: unrelated change, also it's wrong

File: src/detect-bytemath.c [hunk 7]
```diff
@@ -418,7 +418,7 @@ static int DetectByteMathSetup(DetectEngineCtx *de_ctx, Signature *s, const char
         SCFree(nbytes);
     DetectByteMathFree(de_ctx, data);
     return ret;
-}
+ }
```
Comment (victorjulien) [inline src/detect-bytemath.c:421]: unrelated change, also it's wrong

File: src/detect-bytetest.c [hunk 8]
```diff
@@ -714,7 +714,7 @@ static int DetectBytetestSetup(DetectEngineCtx *de_ctx, Signature *s, const char
         SCFree(nbytes);
     DetectBytetestFree(de_ctx, data);
     return ret;
-}
+ }
```
Comment (victorjulien) [inline src/detect-bytetest.c:717]: unrelated change, also it's wrong

File: src/detect-engine-build.c [hunk 9]
```diff
@@ -1624,7 +1624,7 @@ int CreateGroupedPortList(DetectEngineCtx *de_ctx, DetectPort *port_list, Detect
             /* when a group's sigs are added to the joingr, we can free it */
             gr->next = NULL;
             DetectPortFree(de_ctx, gr);
-        /* append */
+            /* append */
```
Comment (victorjulien) [inline src/detect-engine-build.c:1627]: unrelated

File: src/detect-ipproto.c [hunk 10]
```diff
@@ -425,7 +425,7 @@ static int DetectIPProtoSetup(DetectEngineCtx *de_ctx, Signature *s, const char
 
     DetectIPProtoFree(de_ctx, data);
     return -1;
-}
+ }
```
Comment (victorjulien) [inline src/detect-ipproto.c:428]: unrelated change, also it's wrong

File: ebpf/xdp_synproxy_kern.c (hunk 4 above)
Comment (vincentmli) [reply ebpf/xdp_synproxy_kern.c:1]: this is from https://elixir.bootlin.com/linux/latest/source/tools/testing/selftests/bpf/progs/xdp_synproxy_kern.c with modification by removing some code

File: src/detect-byte-extract.c (hunk 5 above)
Comment (vincentmli) [reply src/detect-byte-extract.c:632]: when I send the PR, the test complains code style error, so I followed the clang-format to format the code, it seems these changes come from clang-format. I will see if I can remove them

File: ebpf/xdp_synproxy.c (hunk 3 above)
Comment (vincentmli) [reply ebpf/xdp_synproxy.c:1]: this is from the Linux kernel tree bpf selftests https://elixir.bootlin.com/linux/latest/source/tools/testing/selftests/bpf/xdp_synproxy.c, I did modify it by removing some code from it.

File: ebpf/include/vmlinux/vmlinux_common.h (hunk 2 above)
Comment (vincentmli) [reply ebpf/include/vmlinux/vmlinux_common.h:1]: these header files are from https://github.com/xdp-project/bpf-examples/tree/master/headers/vmlinux with modification, the aim is to include minimum header files for the bpf program, not include full vmlinux.h header file which is too big

**Comment by vincentmli** [comment]
> Interesting work. Some comments and questions inline.
> 
> Big question for me: can you share some results?

what result you expect to see? for the XDP bpf program performance, the original author has good performance result in 
 https://netdevconf.info/0x15/slides/30/Netdev%200x15%20Accelerating%20synproxy%20with%20XDP.pdf, maybe add a reference link to it in doc ?

**Comment by victorjulien** [comment]
> > Interesting work. Some comments and questions inline.
> > Big question for me: can you share some results?
> 
> what result you expect to see? for the XDP bpf program performance, the original author has good performance result in https://netdevconf.info/0x15/slides/30/Netdev%200x15%20Accelerating%20synproxy%20with%20XDP.pdf, maybe add a reference link to it in doc ?

Suricata with and without this, under some kind of stress test.

**Comment by vincentmli** [comment]
> Suricata with and without this, under some kind of stress test.

 I did SYN flooding test with and without this with Suricata, Suricata CPU spikes near 100% without this and enter into emergency mode, with this, Suricata CPU usage is about 6% under SYN flood as normal. or you are concerned if this feature introduce any overhead to Suricata itself in data path and need some other kind of stress test? I am not familiar with Suricata stress test, so any reference documentation on  stress test would be great.

**Comment by vincentmli** [comment]
I updated the commit message with Suricata CPU usage when under SYN flooding attack without and with this feature, the CPU usage difference is from nearly 99% to 0.0%, the is expected since XDP is in early data path before Suricata AF_PACKET, Suricata will not even see the SYN flood packet since XDP already dropped SYN flood attach

with:
```
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                       
   1703 root      20   0  752860 372940 329216 S   0.3   1.1   0:00.32 FM#01                                         
   1699 root      20   0  752860 372940 329216 S   0.0   1.1   0:00.52 Suricata-Main                                 
   1702 root      20   0  752860 372940 329216 S   0.0   1.1   0:00.06 W#01-eno1                                     
   1704 root      20   0  752860 372940 329216 S   0.0   1.1   0:00.00 FR#01                                         
   1705 root      20   0  752860 372940 329216 S   0.0   1.1   0:00.00 CW                                            
   1706 root      20   0  752860 372940 329216 S   0.0   1.1   0:00.00 CS                                            
   1707 root      20   0  752860 372940 329216 S   0.0   1.1   0:00.00 US     
```

without:

```
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                       
   1702 root      20   0  752860 413900 329216 R  99.9   1.3   0:11.12 W#01-eno1                                     
   1703 root      20   0  752860 413900 329216 S   0.7   1.3   0:00.60 FM#01                                         
   1699 root      20   0  752860 413900 329216 S   0.0   1.3   0:00.64 Suricata-Main                                 
   1704 root      20   0  752860 413900 329216 S   0.0   1.3   0:00.00 FR#01                                         
   1705 root      20   0  752860 413900 329216 S   0.0   1.3   0:00.00 CW                                            
   1706 root      20   0  752860 413900 329216 S   0.0   1.3   0:00.01 CS                                            
   1707 root      20   0  752860 413900 329216 S   0.0   1.3   0:00.01 US                                            
```

**Comment by vincentmli** [comment]
@victorjulien any further feedback?

**Comment by catenacyber** [comment]
I think Victor expects a new rebased PR (instead of pushes on the existing one which has changes-requested status)

**Comment by vincentmli** [comment]
> I think Victor expects a new rebased PR (instead of pushes on the existing one which has changes-requested status)

ok, the process works different for other project, I will do rebase then and send new PR.
